docs/meta-intro/intro-to-meta.md

```diff
@@ -0,0 +1,624 @@
+---
+nav_order: 1
+---
+# Introduction to Meta-Programming
+
+During this lab you will be introduced to the basic concepts of meta-programming: compile time execution and templates.
+
+## Manifest Constants
+
+Manifest constants are variables evaluated at compile time.
+They cannot change their value.
+A manifest constant is declared by using the keywork **enum**:
+
+```d
+enum i = 4;      // i is 4 of type int
+enum long l = 3; // l is 3 of type long
+```
+
+Manifest constants are not lvalues, meaning their address cannot be taken. They exist only in the memory of the compiler.
+Declaring a manifest constant that cannot be evaluated at compile time is an error:
+
+```d
+void main()
+{
+    int a = 5;
+    enum b = 5 + 2;   // ok, '5' and '2' are integer literals, known at compile time
+    enum b = a + 2;   // error: 'a' cannot be read at compile time  
+}
+```
+
+To make the above code work, **a** should be declared as an **enum**.
+
+Manifest constants can be seen as compile time variable declarations.
+
+## Compile time function execution (CTFE)
+
+CTFE is a mechanism which allows the compiler to execute functions at compile time.
+There is no special set of the D language necessary to use this feature - whenever a function just depends on compile time known values the D compiler might decide to interpret it during compilation.
```

Comment:
This is not exactly true. The compiler doesn't decide randomly to execute at compile time.